lsp/src/main/kotlin/org/gradle/declarative/lsp/DeclarativeTextDocumentService.kt

```diff
@@ -502,54 +561,35 @@ private class ValueFactoryIndexStore() {
-    //  in which block, because only the ones defined at the top level are available everywhere
-
-    fun factoriesForProperty(fqName: FqName): List<LabelAndInsertText>? = index[fqName]
-
-    private fun indexValueFactories(analysisSchema: AnalysisSchema, type: DataClass, namePrefix: String): Map<FqName, List<LabelAndInsertText>> {
-        val factoryIndex = mutableMapOf<FqName, List<LabelAndInsertText>>()
-
-        type.memberFunctions
-            .filter { it.semantics is FunctionSemantics.Pure && it.returnValueType is DataTypeRef.Name }
-            .forEach {
-                val indexKey = (it.returnValueType as DataTypeRef.Name).fqName
-                val labelAndInsertText = LabelAndInsertText("$namePrefix${computeCompletionLabel(it)}", "$namePrefix${computeCompletionInsertText(it, analysisSchema)}")
-                factoryIndex.merge(indexKey, listOf(labelAndInsertText)) { oldVal, newVal -> oldVal + newVal }
-            }
-
-        val typeRefContext = SchemaTypeRefContext(analysisSchema)
-        type.properties.forEach {
-            when (val propType = typeRefContext.resolveRef(it.valueType)) {
-                is DataClass -> {
-                    val propName = it.name
-                    val propIndex = indexValueFactories(analysisSchema, propType, "$namePrefix${propName}.")
-                    propIndex.forEach { (key, value) -> factoryIndex.merge(key, value) { oldVal, newVal -> oldVal + newVal } }
-                }
-                else -> Unit
+private fun DataType.ParameterizedTypeInstance.TypeArgument.toSimpleName(
+    genericTypeSubstitution: Map<DataType.TypeVariableUsage, DataType> = emptyMap()
+) =
+    when (this) {
+        is DataType.ParameterizedTypeInstance.TypeArgument.ConcreteTypeArgument ->
+            if (type is DataTypeRef.Type && (type as DataTypeRef.Type).dataType is DataType.TypeVariableUsage) {
+                genericTypeSubstitution[(type as DataTypeRef.Type).dataType].toString()
+            } else {
+                this.toString()
             }
-        }
-
-        return factoryIndex
+        is DataType.ParameterizedTypeInstance.TypeArgument.StarProjection -> this.toString()
     }
-}
-
-private data class LabelAndInsertText(val label: String, val insertText: String)
 
-// Extension functions -------------------------------------------------------------------------------------------------
-
-// TODO: this might not be the best way to resolve the type name, but it works for now
-private fun DataTypeRef.toSimpleName(): String = this.toString()
-
-private fun DataParameter.toSignatureLabel() = "${this.name}: ${this.type.toSimpleName()}"
+private fun DataParameter.toSignatureLabel(
+    genericTypeSubstitution: Map<DataType.TypeVariableUsage, DataType> = emptyMap()
+): String = "${this.name}: " +
+        when (this) {
+            is VarargParameter -> "vararg ${type.toSimpleName(genericTypeSubstitution)}"
```

Comment:
💭 In the regular Kotlin function declaration syntax, `vararg` is a prefix of the whole parameter, e.g. `vararg strings: String`. Should we follow that notation here, too?

Reply:
Yup, completely agree, but wasn't straightforward to do, at least not here. Shouldn't we do that in the `VarargParameter`'s code? I mean we know it's a vararg, so why do we point to an array in the `type` field, instead of directly to the type of elements in the array?

Reply:
Fixed